Add JS to make gtag work with Turbolinks

citizen428 opened the pull request:
@roonglit pointed out a problem with our current Google Analytics implementation, namely that it won't work nicely with Turbolinks, since the script in the head won't execute again. 

Resources:

* [Tracking virtual pageviews](https://developers.google.com/analytics/devguides/collection/gtagjs/single-page-applications#tracking_virtual_pageviews)
* [gist](https://gist.github.com/esBeee/545653241530f8f2c2e16371bec56f20#gistcomment-2310777)

citizen428 commented:
@janklimo This is still WIP, waiting for @roonglit to verify this on the project where he initially found the problem.

roonglit commented on app/assets/javascripts/google_analytics.js.erb:
```diff
@@ -0,0 +1,3 @@
+$(document).on('turbolinks:load', function () {
+  gtag('config', '<%= ENV['GOOGLE_ANALYTICS_ID'] %>', { 'page_path': window.location.pathname });
```
I'm avoiding using .erb for assets. how about this?
```
if (typeof gtag === 'undefined') { return; }
  gtag('config', GOOGLE_ANALYTICS_ID, {
    'page_title' : document.title,
    'page_path': window.location.pathname
  });

```

roonglit replied:
we also need to check if gtag is a function or not.

citizen428 replied:
Question: why would you avoid using `erb` for assets? The asset pipeline will always execute the extensions in order, so there's no technical reason not to use it.

Good catch re checking for being a function.

roonglit commented on app/views/layouts/_google_analytics.html.erb:
```diff
@@ -1,11 +1,7 @@
 <script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments)};
   gtag('js', new Date());
```
in order to use the code above, I'm adding the line below:
`var GOOGLE_ANALYTICS_ID = '<%= ENV['GOOGLE_ANALYTICS_ID'] %>';`

citizen428 replied:
Not entirely sure what you mean.

hiattp replied:
In projects based on standard js/jquery as opposed to angular/react I frequently find myself defining a select few global variables to re-use through the assets, e.g. in the application layout (or a partial therein) I'll put something like `var TRUSTBOOKS = {}` in js so that I can load up necessary ENV variables or other important (potentially dev/staging/production-specific configs), mainly just to keep them centralized (as opposed to in js.erb files). then I always know where to look, and js files stay as pure js files in case we want to do something unusual with the asset pipeline.

I don't know if that is relevant here, just something I've done in the past and continue to do now. Maybe it's something we should encourage in the template? Or maybe there is a best practice around this I don't know about 😁 

Anyway not super opinionated about these changes so I'll let you guys hash this out 🙏

citizen428 replied:
@roonglit If you wanna do me a favor it'd be great if you could add to this PR and maybe implement @hiattp's approach (which is pretty similar to yours). Maybe we can also standardize on a name, like e.g. `SETTINGS`.

citizen428 commented on app/views/layouts/application.html.erb:
```diff
@@ -10,7 +10,9 @@
     <%= render partial: 'layouts/favicon' %>
     <link rel='manifest' href='manifest.json'>
-    <%= render partial: 'layouts/google_analytics' %>
+    <% if ENV['GOOGLE_ANALYTICS_ID'].present? %>
```
Mostly because it sort of hid the fact that the partial may not render anything at all. But if the majority of the people agree I can move it back there.

citizen428 replied:
@janklimo @hiattp @roonglit So, what's everyone's preference on this? Keep the conditional in the partial so the layout stays cleaner or be more explicit about the conditions under which th Google Analytics code will be added?

janklimo replied:
@citizen428 I lean towards keeping them in the partial. More in line with _tell, don't ask_ IMO.

citizen428 commented:
Ok, I added a partial called `_js_setup.html.erb` that sets up a global `SETTINGS` object and updated the rest of the code to use it. @hiattp I looked at some other projects to see what you usually do, but I couldn't find any shared state objects like you mentioned in your comments.

citizen428 commented:
Thanks @roonglit for flagging this and @hiattp and @janklimo for the feedback, gonna merge this now.
